manifests/testing/disks-images-provider.yaml.in

```diff
@@ -100,6 +100,10 @@ spec:
         - name: target
           image: {{.DockerPrefix}}/disks-images-provider:{{.DockerTag}}
           imagePullPolicy: Always
+          lifecycle:
+            preStop:
```

Comment:
This looks promising.
A few questions/points:
- On kind provider, we will gave multiple loopback devices pointing to the same image, one from each node. So I wander if it will be correct to poss the exact loopdevice though an env variable (from the entrypoint.sh).
- Will this hook work when we forcibly kill the cluster using `make cluster-down`?

Reply:
Thanks for your quick reply.

> * On kind provider, we will gave multiple loopback devices pointing to the same image, one from each node. So I wander if it will be correct to pass the exact loopdevice though an env variable (from the entrypoint.sh).

I am not sure if my uderstanding is right.
Do you mean we pass loopdevice name (e.g. /dev/loop2) to an env variable via "entrypoint.sh", then in preStop hook, using this env variable to detach the exact loopdevice?

> * Will this hook work when we forcibly kill the cluster using `make cluster-down`?

Good point, I just check it. "make cluster-down" does not trigger preStop hook. We may need to clean up attached loopdevice in cluster-down script.

Reply:
> Thanks for your quick reply.
> 
> > * On kind provider, we will gave multiple loopback devices pointing to the same image, one from each node. So I wander if it will be correct to pass the exact loopdevice though an env variable (from the entrypoint.sh).
> 
> I am not sure if my uderstanding is right.
> Do you mean we pass loopdevice name (e.g. /dev/loop2) to an env variable via "entrypoint.sh", then in preStop hook, using this env variable to detach the exact loopdevice?

Yes. Just an option.
> 
> > * Will this hook work when we forcibly kill the cluster using `make cluster-down`?
> 
> Good point, I just check it. "make cluster-down" does not trigger preStop hook. We may need to clean up attached loopdevice in cluster-down script.

Yes, or we can make the `cluster-clean` a dependency of `cluster-down` in the makefile. If that works, it will be the cleanest solution (I hope).

Reply:
please consider that we sometimes do docker rm --force , or there is power cut, 
i would suggest that there will be a manual flow for local machines, and maybe init script on prow nodes for this, in case its a real issue ?

Reply:
> please consider that we sometimes do docker rm --force , or there is power cut,
> i would suggest that there will be a manual flow for local machines, and maybe init script on prow nodes for this, in case its a real issue ?

One suggestion above was to do it in cluster-clean target, how is this not solving the scenarios you mentioned?

Reply:
it solves, right, btw, it act already when `cluster-deploy` is triggered, so maybe even no need to make `cluster-clean` binding to `cluster-down` ?

Reply:
We should clean on exit, always.

Rethinking this, we can only cleanup the devices the current cluster created, as other clusters may run in parallel, especially on CI.
Therefore, we need a way to share the new created devices. I cluster-clean gracefully cleans up the running cluster, we have the hook. If something else goes bad, we may need another mechanism (this may be complex).
I am good to start with graceful cleaning.

Reply:
power cut wont let you clean on exit, is it ?

Reply:
Hi, @EdDev @oshoval , I updated the patch. Now attached loopdev would be clean up when disks-images-provider is deleted or use "make cluster-down". 

But if the kind cluster is killed via "docker kill", loopdev leak issue would still exist.

Reply:
Great, yep also if there is a power cut, this is why i think it should be done also on cluster-up
(it can know if the files are used for example by holding a lock file, even not on this PR, just my opinion at least).

Thanks

Reply:
@oshoval You repeat the power cut as if this is a fact and not an assumption. Please check it and let us know if this is indeed an issue. I claim it is not.

> But if the kind cluster is killed via "docker kill", loopdev leak issue would still exist.

This scenario is a far edge case and it is the problem of the user who decides to take such actions. I guess this can be handled somehow, but I would not invest in this now unless we see this as a real problem in the field (e.g. CI).

Reply:
> Great, yep also if there is a power cut, this is why i think it should be done also on cluster-up
> (it can know if the files are used for example by holding a lock file, even not on this PR, just my opinion at least).
> 
> Or meanwhile to have a snippet in the readme, that tells what to make sure before cluster-up and what to run ?
> 
> Thanks

I get your idea, and think about it for a while. Here is a problem, if we do it on cluster-up , how we can desicde which attached loopdevs should be detached?

Reply:
its a fact, as dokcer kill is a fact, as Zhlhahaha,
its so trivial that it would happen if the logic is part of a graceful shutdown, that its not need to be proved, this is my opinion at least, can can be handled in future PR
meanwhile i suggest to have a flow in the readme to over come it.

once the PR is merged we can try the shutdown, not in the middle of it imo, there is the right time for it.

about docker kill, imo,
its not an edge case but what we actually do usually in order to clean the machines locally, there was an email thread about it, its the convinient way, and also worked around bugs of cluster-down, so people do use it, so in this case, at least give the user the knowledge to overcome it, else people will waste time, this is again my opinion.

Reply:
> > Great, yep also if there is a power cut, this is why i think it should be done also on cluster-up
> > (it can know if the files are used for example by holding a lock file, even not on this PR, just my opinion at least).
> > Or meanwhile to have a snippet in the readme, that tells what to make sure before cluster-up and what to run ?
> > Thanks
> 
> I get your idea, and think about it for a while. Here is a problem, if we do it on cluster-up , how we can desicde which attached loopdevs should be detached?

If the device creator, creates a linux lock file and hold it, 
that file has the names of the devices it created,
the other cluster on cluster-up, checks what devices it cant kill , nor use, according the locked files, the other ones it cleans

again just a suggestion for the future, hope it make sense or at least a direction

thanks

Reply:
> its a fact, as dokcer kill is a fact, as Zhlhahaha said. (since its a private case of it)

As I already said, that is the user problem and a far edge case, I am not interested to take care of it now. Especially when there is no good solution at the moment.

> i repeated it until it was clear enough that there are other cases,
> its so trivial that it would happen if the logic is part of a graceful shutdown, that its not need to be proved, this is my opinion at least, it can be handled in future PR
> meanwhile i suggest to have a flow in the readme to over come it.

Opinions are not facts. Power cut is not relevant, if you think it is, prove this is preserved over a power cycle.

The existing readme is enough, it says to do `make cluster-down`, but if you think something needs to be added, say exactly where and what to write there as I do not understand.

> once the PR is merged we can try the shutdown, not in the middle of it imo, there is the right time for it.

You do not need this PR to prove a claim like that.

> 
> about docker kill, imo,
> its not an edge case but what we actually do usually in order to clean the machines locally, there was an email thread about it, its the convinient way, and also worked around bugs of cluster-down, so people do use it, so in this case, at least give the user the knowledge to overcome it, else people will waste time, this is again my opinion.

When you will have a solution on how to do it, post a PR. Users should use `make cluster-down`. CI should never use `kill` so I find this fine without it.

Reply:
> > its a fact, as dokcer kill is a fact, as Zhlhahaha said. (since its a private case of it)
> 
> As I already said, that is the user problem and a far edge case, I am not interested to take care of it now. Especially when there is no good solution at the moment.
> 
> > i repeated it until it was clear enough that there are other cases,
> > its so trivial that it would happen if the logic is part of a graceful shutdown, that its not need to be proved, this is my opinion at least, it can be handled in future PR
> > meanwhile i suggest to have a flow in the readme to over come it.
> 
> Opinions are not facts. Power cut is not relevant, if you think it is, prove this is preserved over a power cycle.
> 
> The existing readme is enough, it says to do `make cluster-down`, but if you think something needs to be added, say exactly where and what to write there as I do not understand.

Just where it says `make cluster-down`
i would add an explanation that in case an abrupt shutdown was caused,
the following lines should be run in order to clean all the leftovers devices.
with a warning that in case there are couple clusters alive, it will delete it for them as well.
This will address the need for the user who use brute kill all, i do care for the users that do it, 
because sometimes there are bugs on `cluster-down`, and they used it meanwhile (this what actually happened)

> 
> > once the PR is merged we can try the shutdown, not in the middle of it imo, there is the right time for it.
> 
> You do not need this PR to prove a claim like that.
> 
> > about docker kill, imo,
> > its not an edge case but what we actually do usually in order to clean the machines locally, there was an email thread about it, its the convinient way, and also worked around bugs of cluster-down, so people do use it, so in this case, at least give the user the knowledge to overcome it, else people will waste time, this is again my opinion.
> 
> When you will have a solution on how to do it, post a PR. Users should use `make cluster-down`. CI should never use `kill` so I find this fine without it.

Even on our machines, when there is lab maintenance, power is cutted,
As Zhlhahaha said, docker fill will skip the delete, power cycle is even stronger hence will cause the same. the same thing can be happen on prow, and we will waste time solving it when it happens.

Reply:
> > The existing readme is enough, it says to do `make cluster-down`, but if you think something needs to be added, say exactly where and what to write there as I do not understand.
> 
> Just where it says `make cluster-down`
> i would add an explanation that in case an abrupt shutdown was caused,
> the following lines should be run in order to clean all the leftovers devices.
> with a warning that in case there are couple clusters alive, it will delete it for them as well.
> This will address the need for the user who use brute kill all, i do care for the users that do it,
> because sometimes there are bugs on `cluster-down`, and they used it meanwhile (this what actually happened)

I would not recommend adding anything, it is too complecated and dengarous.
If the authors really wants it, I would just add a small warning that without a graceful cleanup of the cluster, loopback devices may be left on the system.
(and not mention any action, as that is dangerous)

> Since we are not fixing it in this PR anyway, its better to first wait the final version imo.

> Even on our machines, when there is lab maintenance, power is cutted,
> As Zhlhahaha said, docker fill will skip the delete, power cycle is even stronger hence will cause the same. the same thing can be happen on prow, and we will waste time solving it when it happens.
> At least a readme, would help us, to add it in the server boot script or something like that.

You have been feedbacked that power cycle is irrelevant (hinting that it will not be persisted over boot).
I suggest you confirm your claim before you continue with this further.

Reply:
> > > The existing readme is enough, it says to do `make cluster-down`, but if you think something needs to be added, say exactly where and what to write there as I do not understand.
> > 
> > 
> > Just where it says `make cluster-down`
> > i would add an explanation that in case an abrupt shutdown was caused,
> > the following lines should be run in order to clean all the leftovers devices.
> > with a warning that in case there are couple clusters alive, it will delete it for them as well.
> > This will address the need for the user who use brute kill all, i do care for the users that do it,
> > because sometimes there are bugs on `cluster-down`, and they used it meanwhile (this what actually happened)
> 
> I would not recommend adding anything, it is too complecated and dengarous.
> If the authors really wants it, I would just add a small warning that without a graceful cleanup of the cluster, loopback devices may be left on the system.
> (and not mention any action, as that is dangerous)

imo at least with that warning, a simple line that shows how to delete them is good to have.

> 
> > Since we are not fixing it in this PR anyway, its better to first wait the final version imo.
> 
> > Even on our machines, when there is lab maintenance, power is cutted,
> > As Zhlhahaha said, docker fill will skip the delete, power cycle is even stronger hence will cause the same. the same thing can be happen on prow, and we will waste time solving it when it happens.
> > At least a readme, would help us, to add it in the server boot script or something like that.
> 
> You have been feedbacked that power cycle is irrelevant (hinting that it will not be persisted over boot).
> I suggest you confirm your claim before you continue with this further.

missed that feedback, where is it please, how do you know that ?'
anyway i said two things, and one of them was right, (the docker kill one which is exactly as docker rm --force)

Reply:
> imo at least with that warning, a simple line that shows how to delete them is good to have.

The user may destroy other devices who other services need. One example, another kind cluster that runs in parallel.
Giving the example may cause him to use it and harm to himself.
We do not want that responsibility.


> missed that feedback, where is it please, how do you know that ?

I checked.

> anyway i said two things, and one of them was right, (the docker kill one which is exactly as docker rm --force)

You raised it and you have been answered that there is no simple solution to this and that we do not need it now. If you do have a solution, I suggest you post one in a following PR.

Reply:
> > imo at least with that warning, a simple line that shows how to delete them is good to have.
> 
> The user may destroy other devices who other services need. One example, another kind cluster that runs in parallel.
> Giving the example may cause him to use it and harm to himself.
> We do not want that responsibility.

under a warning i think we do, else people would start asking how to do it.

> 
> > missed that feedback, where is it please, how do you know that ?
> 
> I checked.
Thanks. i couldnt guess that.
At least we did sync about what actually can happen and not and decide accordingly.

> 
> > anyway i said two things, and one of them was right, (the docker kill one which is exactly as docker rm --force)
> 
> You raised it and you have been answered that there is no simple solution to this and that we do not need it now. If you do have a solution, I suggest you post one in a following PR.

i purposed one in other thread in this PR, of course not needed for now (wrote that as well)
and the actual happenings will show if we need it in the future.

Reply:
I think this PR should focus exclusively on 1 thing: **cleanup of the loop device on a graceful shutdown of the cluster**.

Reply:
@oshoval @EdDev I add a echo message in disks-images-provider/entrypoint.sh in case attached loopdevice number reached the limit(100), and leave this issue to user.

Reply:
Thank you!